### ISSUE TITLE
Segfault with vm_compute, primitive projections, and opaque constants.

### ISSUE THREAD

**Issue by Zimmi48** [Zimmi48/bugzilla-test#4203]
Note: the issue was created automatically with bugzilla2github tool

# Bugzilla Bug ID: 4203
Date: 2015-04-22 22:58:12 +0200
From: David M. Cooke &lt;<david.m.cooke@gmail.com>&gt;
To:   
CC:   coq-bugs-redist@lists.gforge.inria.fr, @_silene

Last updated: 2015-07-05 19:15:01 +0200

## Bugzilla Comment ID: 10251
Date: 2015-04-22 22:58:12 +0200
From: David M. Cooke &lt;<david.m.cooke@gmail.com>&gt;

The following code segfaults (8.5 branch):

Set Primitive Projections.

Record ops {T:Type} := { is_ok : T -> Prop; constant : T }.
Arguments ops : clear implicits.

Record ops_ok {T} (Ops:ops T) := { constant_ok : is_ok Ops (constant Ops) }.

Definition nat_ops : ops nat := {| is_ok := fun n => n = 1; constant := 1 |}.
Definition nat_ops_ok : ops_ok nat_ops.
Proof.
  split. cbn. apply eq_refl.
Qed.

Eval vm_compute in constant_ok nat_ops nat_ops_ok.

It doesn't segfault if any of these are done:
* using cbv instead of vm_compute
* removing the 'Set Primitive Projections' line
* finishing the definition of nat_ops_ok with Defined instead of Qed.

Both the native code and bytecode versions of coq segfault.

For what it's worth, if native_compute is used instead, coq throws a 'Stack overflow' exception (with or without 'Set Primitive Projections').

## Bugzilla Comment ID: 10262
Date: 2015-04-26 23:32:50 +0200
From: @_silene

The segfault happens at the following point:

      Instruct(MAKESWITCHBLOCK) {
	print_instr("MAKESWITCHBLOCK");
	*--sp = accu;
	accu = Field(accu,1);
	switch (Tag_val(accu)) {

In other words, the first field of the Accumulator block is not a pointer to a block (representing the data about the opaque constant, I guess). In case it rings a bell, its value is 1024.

## Bugzilla Comment ID: 10267
Date: 2015-04-27 09:39:39 +0200
From: @_silene

Since the native_compute stack overflow is possibly unrelated, I have opened a new report for it: bug #4210.

## Bugzilla Comment ID: 10447
Date: 2015-06-09 14:19:53 +0200
From: @_maximedenes

#4210 is now fixed and was indeed unrelated. The present example now works fine with native_compute but still segfaults with vm_compute.

## Bugzilla Comment ID: 10553
Date: 2015-07-05 19:15:01 +0200
From: @_maximedenes

Fixed in v8.5 branch. Thanks for reporting!
